soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp

```diff
@@ -0,0 +1,1046 @@
+}
+inline ImVec4 GetColorFromCVar(std::string cvarname) {
+    float R = CVar_GetFloat((cvarname + "R").c_str(), 255);
+    float G = CVar_GetFloat((cvarname + "G").c_str(), 255);
+    float B = CVar_GetFloat((cvarname + "B").c_str(), 255);
+    float A = CVar_GetFloat((cvarname + "A").c_str(), 255);
+    return ImVec4(R, G, B, A);
+}
+inline Color_RGBA8 ImVec4_to_Color_RGBA8(ImVec4& color) {
+    Color_RGBA8 ret;
+    ret.r = color.x * 255.0f;
+    ret.g = color.y * 255.0f;
+    ret.b = color.z * 255.0f;
+    ret.a = color.w * 255.0f;
+    return ret;
+}
+inline ImVec4 Color_RGBA8_to_ImVec4(Color_RGBA8& color) {
+    ImVec4 ret;
+    ret.x = color.r / 255.0f;
+    ret.y = color.g / 255.0f;
+    ret.z = color.b / 255.0f;
+    ret.w = color.a;
+    return ret;
+}
+
+void ImGuiDrawTwoColorPickerSection(const char* text, const char* cvarMainName, const char* cvarExtraName,
+                                 Color_RGBA8& main_color, Color_RGBA8& extra_color, Color_RGBA8& main_default_color,
+                                 Color_RGBA8& extra_default_color, const char* cvarHideName) {
+    Color_RGBA8 cvarMainColor = CVar_GetRGBA(cvarMainName, main_default_color);
+    Color_RGBA8 cvarExtraColor = CVar_GetRGBA(cvarExtraName, extra_default_color);
+    main_color = cvarMainColor;
+    extra_color = cvarExtraColor;
+
+    if (ImGui::CollapsingHeader(text)) {
+        if (*cvarHideName != '\0') {
+            std::string label = cvarHideName;
+            label += "##Hidden";
+            ImGui::PushID(label.c_str());
+            UIWidgets::EnhancementCheckbox("Hidden", cvarHideName, false,
+                                           "When active, checks will hide by default when updated to this state. Can "
+                                           "be overriden with the \"Show Hidden Items\" option.");
+            ImGui::PopID();
+        }
+        if (ImGui::BeginTable(text, 2, ImGuiTableFlags_BordersH | ImGuiTableFlags_BordersV | ImGuiTableFlags_Hideable)) {
+            ImGui::TableNextRow();
+            ImGui::TableNextColumn();
+            ImGui::PushItemWidth(ImGui::GetContentRegionAvail().x);
+            if (UIWidgets::EnhancementColor("Check", cvarMainName, Color_RGBA8_to_ImVec4(main_color), Color_RGBA8_to_ImVec4(main_default_color))) {
+                main_color = CVar_GetRGBA(cvarMainName, main_default_color);
+            };
+            ImGui::PopItemWidth();
+            
+            ImGui::TableNextColumn();
+            ImGui::AlignTextToFramePadding();
+            ImGui::PushItemWidth(ImGui::GetContentRegionAvail().x);
+            if (UIWidgets::EnhancementColor("Details", cvarExtraName, Color_RGBA8_to_ImVec4(extra_color),
+                                            Color_RGBA8_to_ImVec4(extra_default_color))) {
+                main_color = CVar_GetRGBA(cvarExtraName, extra_default_color);
+            }
+            ImGui::PopItemWidth();
+
+            ImGui::EndTable();
+        }
+    }
+}
```

Comment:
i feel like these exist in other places already, could this be done without duplicated code?

Reply:
The rainbow code is a modification of Cosmetic Editor's code for rainbows. With a recent refactor of it from develop, I wasn't able to use Cosmetic Editor directly any more. I can look at refactoring both Cosmetic Editor and Check Tracker to both use a new standard rainbow generator? Or take a second crack at jamming in my UI code into the Cosmetics?

The colour CVar save/load can also be refactored to just use the CVar_GetRGBA directly instead of the floats. It was only there to avoid having to rewrite the background colour/opacity. 

I couldn't find a standard implementation of the colour conversion method between RGBA8 and ImVec4 that the ImGui Widgets use. I can directly inline the code if the method is a problem.

Similarly, I wasn't able to see how I could use Cosmetic Editor's grid drawing to my advantage here in the check tracker options. I'll take another look for getting it to work and/or look to see if there's any other places

Reply:
i figure worst case scenario just adding a todo and making an issue to remove the copypasta could work, but hopefully you can figure something out!

Reply:
Commit e9e0a36add27b24573f62fd1bf7483caee3c7318 fixes a bug in UIWidgets where alpha-enabled EnhancementColor was practically useless, and 47af3dfd35a4e72310de48a5fbe8b79a12214542 refactors to get rid of the inline methods. Although, it doesn't seem much better. I would argue that Color_RGBA8_to_ImVec4 kept the code cleaner, and I couldn't find an alternative anywhere. 

As for the other points on the Cosmetic Editor, I still don't see a way of easily getting the two code-bases to cooperate. The two colour picker cannot be replaced by DrawCosmeticGroup, and the rainbow with CosmeticsUpdateTick, without major rework two both modules.

Reply:
let's make a follow up issue for it, combining the codebases feels out of scope for this

as for https://github.com/HarbourMasters/Shipwright/commit/e9e0a36add27b24573f62fd1bf7483caee3c7318, i'm guessing you mean it wasn't being used before and the incorrect logic wasn't noticed?